src/theory/quantifiers/ho_trigger.cpp

```diff
@@ -0,0 +1,343 @@
+bool HigherOrderTrigger::sendInstantiation( InstMatch& m, unsigned var_index ) {
+  if( var_index==d_ho_var_list.size() ){
+    // we now have an instantiation to try
+    return d_quantEngine->addInstantiation( d_f, m );
+  }else{
+    Node var = d_ho_var_list[var_index];
+    unsigned vnum = var.getAttribute(InstVarNumAttribute());
+    Assert( vnum<m.d_vals.size() );
+    Node value = m.d_vals[vnum];
+    Assert( d_lchildren[vnum][0] == value );
+    Assert( d_ho_var_bvl.find( var )!=d_ho_var_bvl.end() );
+
+    // now, recurse on arguments to enumerate equivalent matching lambda expressions
+    bool ret = sendInstantiationArg( m, var_index, vnum, 0, d_ho_var_bvl[var], false );
+
+    //reset the value
+    m.d_vals[vnum] = value;
+    
+    return ret;
+  }
+}
+
+bool HigherOrderTrigger::sendInstantiationArg( InstMatch& m, unsigned var_index, unsigned vnum, unsigned arg_index,
+                                               Node lbvl, bool arg_changed ) {
+  
+  if( arg_index==lbvl.getNumChildren() ){
+    // construct the lambda
+    if( arg_changed ){
+      Node body = NodeManager::currentNM()->mkNode( kind::APPLY_UF, d_lchildren[vnum] );
+      Node lam = NodeManager::currentNM()->mkNode( kind::LAMBDA, lbvl, body );
+      m.d_vals[vnum] = lam;
+      Trace("ho-unif-debug2") << "  try " << vnum << " -> " << lam << std::endl;
+    }
+    return sendInstantiation( m, var_index+1 );
+  }else{
+    std::map< unsigned, unsigned >::iterator itr = d_arg_to_arg_rep[vnum].find( arg_index );
+    unsigned rindex = itr!=d_arg_to_arg_rep[vnum].end() ? itr->second : arg_index;
+    std::map< unsigned, std::vector< Node > >::iterator itv = d_arg_vector[vnum].find( rindex );
+    Assert( itv!=d_arg_vector[vnum].end() );
+    Node prev = lbvl[arg_index];
+    bool ret = false;
+    // try each argument in the vector
+    for( unsigned i=0; i<itv->second.size(); i++ ){
+      bool new_arg_changed = arg_changed || prev!=itv->second[i];
+      d_lchildren[vnum][arg_index+1] = itv->second[i];
+      if( sendInstantiationArg( m, var_index, vnum, arg_index+1, lbvl, new_arg_changed ) ){
+        ret = true;
+        break;
+      }
+    }
+    // clean up
+    d_lchildren[vnum][arg_index+1] = prev;
+    return ret;
+  }
+}
+
+int HigherOrderTrigger::addHoTypeMatchPredicateLemmas() {
+  unsigned numLemmas = 0;
+  if( !d_ho_var_types.empty() ){
+    // this forces expansion of APPLY_UF terms to curried HO_APPLY chains
+    for( unsigned j=0; j<d_quantEngine->getTermDatabase()->getNumOperators(); j++ ){
+      Node f = d_quantEngine->getTermDatabase()->getOperator( j );
+      if( f.isVar() ){
+        TypeNode tn = f.getType();
+        if( d_ho_var_types.find( tn )!=d_ho_var_types.end() ){
+          Node u = d_quantEngine->getTermUtil()->getHoTypeMatchPredicate( tn );
+          Node au = NodeManager::currentNM()->mkNode( kind::APPLY_UF, u, f );
```

Comment:
Don't call NodeManager::currentNM() inside the for

Reply:
We should revisit whether this is a strict guideline. You could actually argue there are other things (e.g. getting the getTermUtil() on each iteration) that are similar. I'm wondering if this actually impacts performance.

Reply:
I imagine that if ``size()`` impacts performance so will these calls. Anyway I was just trying to go with the flow.